### PR TITLE
Implement bot manager

### LINKED ISSUE #360: Add bots
- [ ] Create app
- [ ] Add N static bots
- [ ] Add realistic behavior

### PR DESCRIPTION
**Pull request by agustinesco**
Closes #360 

Now whenever you enter a match there'll be always the maximum amount of player, the diffence being that the arena will communicate with a different app through a get request to start a bot to control the missing players

in this iteration the logic will be as simple as possible, move in a random direction and try to attach each 300ms 

The communication will be as follows (You can also find this diagram on the readme file): 
```mermaid
sequenceDiagram

    box  Arena
    participant Arena.GameSocketHandler
    participant GameLauncher
    participant GameUpdater
    end
    box  Bot Manager
    participant Endpoint
    participant BotSupervisor
    participant BotManager.GameSocketHandler
    end

    GameLauncher->>+GameLauncher:start_by_timeout
    GameLauncher->>GameLauncher:init_bots(missing_clients)
    GameLauncher->>GameUpdater:start(clients ++ bots_clients) 
    GameUpdater->>GameLauncher:{:ok, game_state} 
    GameLauncher->>-Endpoint:get(join/client_id/game_pid)
    Endpoint->>BotSupervisor:spawn_bot(client_id, game_pid)
    BotSupervisor->>BotManager.GameSocketHandler:init(client_id, game_pid)
    BotManager.GameSocketHandler->>Arena.GameSocketHandler:connect(client_id, game_pid)
    
    loop Every 300ms
        BotManager.GameSocketHandler->>Arena.GameSocketHandler: do_action()
        Arena.GameSocketHandler->>GameUpdater:do_action(player_id, game_id)
    end
    
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -52,9 +52,17 @@ generate-game-client-protos:
 		--proto_path=apps/serialization \
 		messages.proto
 
+
```
Comment (manucamejo) [inline Makefile:55]: remove this line

File: apps/arena/lib/arena/game_updater.ex [hunk 2]
```diff
@@ -33,10 +33,10 @@ defmodule Arena.GameUpdater do
   # END API
   ##########################
 
-  def init(%{clients: clients}) do
-    game_id = self() |> :erlang.term_to_binary() |> Base58.encode()
+  def init(%{clients: clients, missing_clients: missing_clients}) do
```
Comment (manucamejo) [inline apps/arena/lib/arena/game_updater.ex:36]: Wait! The game shouldn't know that some clients are bots. This logic should live inside the launcher.
Regarding the core of the game, all the clients are equal.

Please move it to the `game_launcher` file.

**Review by manucamejo — CHANGES_REQUESTED** [review]

File: apps/arena/lib/arena/game_updater.ex (hunk 2 above)
Comment (agustinesco) [reply apps/arena/lib/arena/game_updater.ex]: you're right! the game should only know about the game state and how to change it, moved the bot initialization to the game launcher as suggested, i also updated the links with the bots names and updated the mermaid diagrams

File: Makefile (hunk 1 above)
Comment (agustinesco) [reply Makefile]: done

File: apps/arena/lib/arena/game_launcher.ex [hunk 3]
```diff
@@ -41,17 +42,28 @@ defmodule Arena.GameLauncher do
...
+        nil
     end
 
     {:noreply, state}
   end
 
   def handle_info(:start_game, state) do
-    {game_clients, remaining_clients} = Enum.split(state.clients, @clients_needed)
+    {game_clients, remaining_clients} =
+      Enum.split(state.clients, @clients_needed)
```
Comment (AminArria) [inline apps/arena/lib/arena/game_launcher.ex:61]: Check your formatter, you didn't change anything here and it still got formatted and even more its not following our rules (line_length 120)

File: apps/bot_manager/mix.exs [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  # Run "mix help deps" to learn about dependencies.
+  defp deps do
+    [
+      # {:dep_from_hexpm, "~> 0.3.0"},
+      # {:dep_from_git, git: "https://github.com/elixir-lang/my_dep.git", tag: "0.1.0"},
+      # {:sibling_app_in_umbrella, in_umbrella: true}
+      # server
+      {:plug_cowboy, "~> 2.2"},
+      # json encode/decode
+      {:poison, "~> 4.0.1"},
```
Comment (AminArria) [inline apps/bot_manager/mix.exs:35]: Use `jason` we already use it in a bunch of other apps and better to be consistent

File: apps/bot_manager/mix.exs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+  def project do
+    [
+      app: :bot_manager,
+      version: "0.1.0",
+      build_path: "../../_build",
+      config_path: "../../config/config.exs",
+      deps_path: "../../deps",
+      lockfile: "../../mix.lock",
+      elixir: "~> 1.14",
```
Comment (AminArria) [inline apps/bot_manager/mix.exs:12]: Go with `1.16`

Comment (agustinesco) [reply apps/bot_manager/mix.exs]: Done!

File: apps/bot_manager/mix.exs (hunk 4 above)
Comment (agustinesco) [reply apps/bot_manager/mix.exs]: Done!

File: apps/arena/lib/arena/game_launcher.ex (hunk 3 above)
Comment (agustinesco) [reply apps/arena/lib/arena/game_launcher.ex:54]: Hmmm i think i'm using the same formatter as everybody :c i'll restore this but i may need some help with this

File: apps/bot_manager/lib/game_socket_handler.ex [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+                 y: y
+               }
+             }
+           }},
+        timestamp: timestamp
+      })
+
+    WebSockex.cast(self(), {:send, {:binary, game_action}})
+
+    Process.send_after(self(), :attack, 300, [])
```
Comment (manucamejo) [inline apps/bot_manager/lib/game_socket_handler.ex:81]: I think that they are going to try to move and attack at the same time every time.
Because you have a `300` here and `@message_delay_ms`at line `35`.

**Review by manucamejo — CHANGES_REQUESTED** [review]

File: config/dev.exs [hunk 7]
```diff
@@ -155,3 +155,9 @@ config :game_client, dev_routes: true
 ##############################
 # App configuration: gateway #
 ##############################
+
+##################################
+# App configuration: bot_manager #
+##################################
+
+config :bot_manager, :end_point, port: [port: 5000]
```
Comment (lotuuu) [inline config/dev.exs:163]: I failed to run the app initially because port 5000 is occupied in mac by the control center unless disabled: https://nono.ma/port-5000-used-by-control-center-in-macos-controlce. To avoid anyone else from running into this issue, I think we could change this port

**Review by lotuuu — CHANGES_REQUESTED** [review]
Bots are added fine but they don't seem to be moving at all 🤔

Uploading Screen Recording 2024-03-22 at 10.41.04.mov…

**Comment by lotuuu** [comment]
Also @agustinesco how would you feel about adding some info logs on bot creation? I think it's a nice utility addition

**Comment by lotuuu** [comment]
Also, let's work a little bit on the names of the bots as we discussed! Nothing too fancy, just using some random names from a predefined list, then we can iterate on it later.

File: apps/bot_manager/lib/game_socket_handler.ex [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+      })
+
+    WebSockex.cast(self(), {:send, {:binary, game_action}})
+
+    {:ok, state}
+  end
+
+  def handle_info(:attack, state) do
+    timestamp = DateTime.utc_now() |> DateTime.to_unix(:millisecond)
+    {x, y} = create_random_movement()
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/bot_manager/lib/game_socket_handler.ex:61]: yeah that name seems more suitable! changed

File: apps/bot_manager/lib/game_socket_handler.ex [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+  def handle_frame(_frame, state) do
+    {:ok, state}
+  end
+
+  def handle_info(:move, state) do
+    Process.send_after(
+      self(),
+      :move,
+      @message_delay_ms
+    )
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/bot_manager/lib/game_socket_handler.ex:36]: sure thing

File: apps/bot_manager/lib/game_socket_handler.ex [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+      })
+
+    WebSockex.cast(self(), {:send, {:binary, game_action}})
+
+    Process.send_after(self(), :attack, 300, [])
+    {:ok, state}
+  end
+
+  def handle_cast({:send, {_type, _msg} = frame}, state) do
+    # Logger.info("Sending frame with payload: #{msg}")
```
[the remark this answers is not in the document]
Comment (agustinesco) [reply apps/bot_manager/lib/game_socket_handler.ex:86]: Deleted

**Comment by agustinesco** [comment]
> Also @agustinesco how would you feel about adding some info logs on bot creation? I think it's a nice utility addition

Yeah seems like a good idea! in fact when i was creating the bots i had some inspects with this information so i see it's utility

File: config/dev.exs (hunk 7 above)
Comment (agustinesco) [reply config/dev.exs:163]: Yeah i was thinking the same, also toxi proxy is using this port so a change was needed. Moved to port 4003

File: apps/arena/lib/arena/game_launcher.ex [hunk 11]
```diff
@@ -86,11 +90,31 @@ defmodule Arena.GameLauncher do
...
   end
 
-  defp get_bot_clients(missing_clients) do
-    Enum.map(1..missing_clients, fn _ ->
+  defp get_bot_clients(missing_clients, game_config) do
+    bot_names = [
+      "TheBlackSwordman",
+      "SlashJava",
+      "SteelBallRun",
+      "Bot h4ck",
```
Comment (lotuuu) [inline apps/arena/lib/arena/game_launcher.ex:98]: Appreciate the meme but let's change this so it doesn't look like a bug 😂 Same with "Bot Name"

File: config/dev.exs [hunk 12]
```diff
@@ -160,4 +160,4 @@ config :game_client, dev_routes: true
 # App configuration: bot_manager #
 ##################################
 
-config :bot_manager, :end_point, port: [port: 5000]
+config :bot_manager, :end_point, port: [port: 4003]
```
Comment (lotuuu) [inline config/dev.exs:163]: Let's use the BOT_MANAGER_PORT env var you used in GameLauncher

**Review by lotuuu — CHANGES_REQUESTED** [review]

File: apps/arena/lib/arena/game_launcher.ex (hunk 11 above)
Comment (agustinesco) [reply apps/arena/lib/arena/game_launcher.ex]: done! 😭

File: config/dev.exs (hunk 12 above)
Comment (agustinesco) [reply config/dev.exs]: done

**Review by lotuuu — APPROVED** [review]

File: apps/arena/lib/arena/game_launcher.ex [hunk 13]
```diff
@@ -63,11 +75,60 @@ defmodule Arena.GameLauncher do
...
   defp maybe_make_batch_start_at([], _) do
     System.monotonic_time(:millisecond)
   end
 
   defp maybe_make_batch_start_at([_ | _], batch_start_at) do
     batch_start_at
   end
+
+  defp get_bot_clients(missing_clients, game_config) do
+    bot_names = [
```
Comment (AminArria) [inline apps/arena/lib/arena/game_launcher.ex:94]: A static list like this should go to a module variable `@bot_names`

File: apps/arena/lib/arena/game_updater.ex [hunk 14]
```diff
@@ -33,9 +33,9 @@ defmodule Arena.GameUpdater do
   # END API
   ##########################
 
-  def init(%{clients: clients}) do
+  def init(%{clients: clients, game_config: game_config}) do
```
Comment (AminArria) [inline apps/arena/lib/arena/game_updater.ex:36]: Why this change? Does it make sense for matchmaking to decide the config of a match?

Also, if we do separate matchmaking into its own app/server then this makes the communication more complex as we need to pass a configuration

**Review by AminArria — CHANGES_REQUESTED** [review]

File: apps/arena/lib/arena/game_launcher.ex (hunk 13 above)
Comment (agustinesco) [reply apps/arena/lib/arena/game_launcher.ex:94]: Changed!

File: apps/arena/lib/arena/game_updater.ex (hunk 14 above)
Comment (agustinesco) [reply apps/arena/lib/arena/game_updater.ex:36]: no it doesn't it was something to randomize the character from the bots, i'll restore this and try another approach, maybe hardcoding since this is a first iteration

**Review by AminArria — DISMISSED** [review]

**Review by lotuuu — APPROVED** [review]

**Review by manucamejo — APPROVED** [review]
